fifteen_min/src/find_amenities.rs

```diff
@@ -0,0 +1,145 @@
+use map_gui::tools::{ChooseSomething, ColorLegend};
+use map_gui::ID;
+use map_model::AmenityType;
+use widgetry::{
+    Cached, Choice, Color, Drawable, EventCtx, GfxCtx, HorizontalAlignment, Key, Line, Panel,
+    SimpleState, State, Transition, VerticalAlignment, Widget,
+};
+
+use crate::isochrone::{Isochrone, Options};
+use crate::viewer::{draw_star, draw_unwalkable_roads, HoverKey, HoverOnBuilding};
+use crate::App;
+
+/// Calculate isochrones around each amenity on a map and merge them together using the min value
+pub struct FindAmenity {
+    options: Options,
+}
+
+impl FindAmenity {
+    pub fn new(ctx: &mut EventCtx, options: Options) -> Box<dyn State<App>> {
+        ChooseSomething::new(
+            ctx,
+            "Choose an amenity",
+            AmenityType::all()
+                .into_iter()
+                .map(|at| Choice::new(at.to_string(), at))
+                .collect(),
+            Box::new(move |at, ctx, app| {
+                let multi_isochrone = create_multi_isochrone(ctx, app, at, options.clone());
+                return Transition::Push(Results::new(ctx, app, multi_isochrone, at));
+            }),
+        )
+    }
+}
+
+/// For every one of the requested amenity on the map, draw an isochrone
+fn create_multi_isochrone(
+    ctx: &mut EventCtx,
+    app: &App,
+    category: AmenityType,
+    options: Options,
+) -> Isochrone {
+    let map = &app.map;
+    // For a category, find all matching stores
+    let mut stores = Vec::new();
+    for b in map.all_buildings() {
+        if b.has_amenity(category) {
+            stores.push(b.id);
+        }
+    }
+    Isochrone::new(ctx, app, stores, options.clone())
+}
+
+struct Results {
+    draw: Drawable,
+    isochrone: Isochrone,
+    hovering_on_bldg: Cached<HoverKey, HoverOnBuilding>,
+    // TODO Can't use Cached due to a double borrow
+    hovering_on_category: Option<(AmenityType, Drawable)>,
+    draw_unwalkable_roads: Drawable,
+}
+
+impl Results {
+    fn new(
+        ctx: &mut EventCtx,
+        app: &App,
+        isochrone: Isochrone,
+        category: AmenityType,
+    ) -> Box<dyn State<App>> {
+        let panel = Panel::new(Widget::col(vec![
+            Line(format!("{} within 15 minutes", category))
+                .small_heading()
+                .into_widget(ctx),
+            ctx.style()
```

Comment:
Similar to the "find your walkable home" and other panels, I think a "close" button would be better than a "back" button.

<img width="1123" alt="Screen Shot 2021-05-12 at 3 28 50 PM" src="https://user-images.githubusercontent.com/217057/118051989-d2224800-b336-11eb-8ff0-5372d4e9a467.png">

Reply:
Thank you for taking a look. I have certainly learned a lot playing around with this part of the app. I do think there is overlap between this and the find a home feature, they are both highlighting areas serviced by certain amenities. In the find a home case it can be multiple amenities but they are not currently distigueshed by walking distance or by amenity. In the find by amenity case it is a single amenity with each location shown and a distance away from any one point shown. My use case is what areas are well covered by say a school and which areas are lacking coverage (with the idea to build more tooling around where would the most effective placement be of a school to meet certain objectives, census and other data merged in could really help here). This could be extended to show multiple amenities like the find a home does now, however from a data vis perspective it is quite noisy (maybe when you plot a star or whatever icon represents each icon it could be a different color per amenity?).

Thank you for pointing out the perfromance difference, it was harder to tell the difference on my machine. Both are using the same path function which has been modified so its good to get to the root of it. For the find by amenity feature I was calcing the duration between a point and all amenities first, finding the minimum (closest amenity) and then determining the path for those two points. On the original home screen when you select a point it was also looking up the duration first then determining a path but you dont need that first lookup so I put an if statement to say if you only have one point we know thats the minimum (ie the building to draw the path to). It should be fixed now, let me know how it goes on your machine. The calcing/drawing is still a bit slow in the find by amenity so its probably worth a more clever way to calc this (maybe there is a way to filter just based on the selected amenity before calling times_from_buildings(), which I believe calcs to the whole map?

I agree on the back button and put that in. I think some feature consolidation can eventually happen as we figure out use cases better for sure. Thanks again!

Reply:
Thanks for explaining, and now I remember your talking about this use case previously ("finding candidate areas to add a school"). I feel like it's different enough from the "find a place to live" use case that having some separate UI like this totally makes sense.

Reply:
I think after this exploration / prototype phase of adding new features, we're going to back up, think about the different use cases, and try some proper UI design for the 15m tool

Reply:
I added the escape button but it seems every time I hit the button it quits my application instead of going back. Any idea why?

Reply:
If you check the terminal, it's crashing: `thread 'main' panicked at 'internal error: entered unreachable code', fifteen_min/src/find_amenities.rs:106:18`
It's in `on_click`:

```
        match x {
            "Back" => Transition::Pop,
            _ => unreachable!(),
        }
```

The change to use the "close" button is the issue. It returns the string "close" as its action, not "Back". This is one of the cases where making panels return an enum of all possible button clicks would be nicer than strings, but that's a hard thing to implement without making every single panel have lots of extra code

Reply:
Ah yes I see what is happening now, I went through button_styles and it makese more sense now. I will just change the code to look out for close.